### DISCUSSION TITLE
json metadata storage of electrical stimulation ?

### DISCUSSION THREAD

**Discussion by etiennedemontalivet** [NeurodataWithoutBorders/helpdesk#70] [Q&A]
Hi,

I am working on experiments that include 2 devices : one to record 20 ECoG channels and another to stimulate through a Functional Electro-Stimulation device that has 10 channels. During each experiment, a specific and unique pattern is used per movement for the electro-stimulation. For instance to generate hand-opening, only the first 3 electrodes are used at specific frequencies. 

**I would like to stored this metadata about the electro-stimulation, but I don't know where in the NWB format.** So far, I though about:
* adding the json data to `NWBFile.protocol`: workaround but `protocol` is suppose to be a string
* adding an `AbstractFeatureSeries` that contains the metadata for each timestamps of the stim data --> memory inefficient...
* adding columns in the electrodes `DynamicTable` to encode the meta data:  
    doable, but:
    * electrodes from `device1` will only have `nan` in these columns
    * information would be hard to read from this "electrode point-of-view"
    * metadata would be stored "per-electrode" and I am looking for something more "movement-related" if possible.
 
 Here is how the metadata looks like:
``` json
{
    "movement_1": {
        "elec": [0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0],
        "freq": 1500,
        "pulses": 3,
        "duration": 500
    },
    "movement_2": { ... }
}
```

Any help would be very much appreciated!

**Comment by CodyCBakerPhD** [comment]
> adding the json data to NWBFile.protocol: workaround but protocol is suppose to be a string

Like `session_description` (what's different about this recording session from the rest of your experiment?) or `experiment_description` (high-level human readable text description of experiment), the task protocol is a text-only descriptor of lower level details about your experiment design -it's essentially what you would write in the 'methods' section of the formal publication related to the data.


> adding columns in the electrodes DynamicTable to encode the meta data:
doable, but:
electrodes from device1 will only have nan in these columns
information would be hard to read from this "electrode point-of-view"
metadata would be stored "per-electrode" and I am looking for something more "movement-related" if possible.

The table approach sounds like the best option to me for recording the metadata about each movement-specific stimulation

Specifically, I would suggest forming a `TimeIntervals` table ([tutorial](https://pynwb.readthedocs.io/en/stable/tutorials/general/plot_timeintervals.html#adding-time-intervals-to-a-nwbfile)) whose `start_time` is the time of stimulation, `stop_time` is the `start_time` + `duration`. Then you'd also have that extra electrodes column, in which I would record the index of the electrodes (from the electrodes table) used to perform the stimulation. Now, if `freq` and `pulses` never change over all your stimulations or stimulation types, then you could either mention these hyperparameters in the description of the table itself, or it would also be perfectly fine to add them as columns and repeat the same value for each stimulation (it's going to be such a comparatively small amount of data in the end; especially if you compress the column). This indicates that in alternative experiments, this 'could have' changed per stimulation

Another detail I will mention is a growing Best Practice in the community - attach the scientific unit to the end of each column name on the table; e.g., `pulses_in_Hz`, `duration_in_seconds` to communicate that detail in a more direct manner (the `description` of the column should still mention it, but not everyone reads those carefully)

**Comment by etiennedemontalivet** [comment]
Thank you for your quick answer!
You are right, I was looking for a way to avoid data duplication, but at the end it's not a lot of data, so I agree that this is better to embed the stimulation metadata directly with every stimulation, as columns of a `TimeIntervals` 👍 
And thanks for the best practice tip :)
